### PR TITLE
Improve getEffectiveBalances()

### PR DESCRIPTION
**Pull request by twoeths**
**Motivation**

Looping through persistent-ts is faster than persistent-merkle-tree

**Description**

+ Also we can use `epochCtx.currentShuffling.activeIndices` to know if a validator is active for currentShuffling epoch or not
+ The benchmark showed that it's a 5x improvement

### CODE REVIEW

File: packages/beacon-state-transition/src/util/balance.ts [hunk 1]
```diff
@@ -52,18 +50,22 @@ export function decreaseBalance(
...
-  const effectiveBalances: number[] = [];
-  const validators = readonlyValuesListOfLeafNodeStruct(justifiedState.validators);
-  for (let i = 0, len = validators.length; i < len; i++) {
-    const validator = validators[i];
-    effectiveBalances.push(isActiveValidator(validator, justifiedEpoch) ? validator.effectiveBalance : 0);
+  const {activeIndices} = justifiedState.currentShuffling;
+  // 5x faster than using readonlyValuesListOfLeafNodeStruct
+  const effectiveBalances = justifiedState.effectiveBalances.toArray();
+  let j = 0;
+  for (let i = 0; i < effectiveBalances.length; i++) {
```
Comment (dapplion) [inline packages/beacon-state-transition/src/util/balance.ts]: If you look at the benchmarks

```
MutableVector 250000 iterate all - toArray() + loop | 8.0552 ms/op | 8.9640 ms/op | 0.90
MutableVector 250000 iterate all - get(i) | 3.4577 ms/op | 3.8777 ms/op | 0.89
```
it's actually cheaper to iterate once with individual get(i) ops. Try creating an empty array and push after each get(i)

Comment (twoeths) [reply packages/beacon-state-transition/src/util/balance.ts]: but we need the number[] in the end, it's convenient to create this way and just update to 0 if it's not an active validator

Comment (twoeths) [reply packages/beacon-state-transition/src/util/balance.ts]: also this needs to address https://github.com/ChainSafe/lodestar/issues/2747#issuecomment-879568043

Comment (twoeths) [reply packages/beacon-state-transition/src/util/balance.ts]: @dapplion in the end I switched to use `persistentTS.get(i)`

File: packages/beacon-state-transition/test/perf/util/balance.test.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import {itBench} from "@dapplion/benchmark";
+import {generatePerfTestCachedStatePhase0, perfStateId} from "../util";
+import {State} from "../types";
+import {getEffectiveBalances} from "../../../src/util";
+
+describe("getEffectiveBalances", () => {
+  itBench<State, State>({
+    id: `getEffectiveBalances - ${perfStateId}`,
```
Comment (dapplion) [inline packages/beacon-state-transition/test/perf/util/balance.test.ts:8]: Use `runsFactor` option to scale down the 100 for loop iterations
https://github.com/dapplion/benchmark/blob/8f2fffda28262e08a446d8a8284d1d2235dc479f/src/types.ts#L40

Comment (twoeths) [reply packages/beacon-state-transition/test/perf/util/balance.test.ts:8]: @dapplion I use `runsFactor: 100` 8 lines below, is that correct?

Comment (dapplion) [reply packages/beacon-state-transition/test/perf/util/balance.test.ts:8]: Oh I'm blind, didn't see it hahaha. Yes it's correct :+1:

**Review by dapplion — APPROVED** [review]

**Pull request merged**